# Imports section

In [98]:

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn import svm
from sklearn import datasets
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import Perceptron

# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

In [7]:
slime = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
slime.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [8]:
slime.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

In [110]:
X = slime[["Temperature °C","Mols KCL"]]
y = slime["Size nm^3"]
X_train, X_test, Y_train, Y_test=train_test_split(X,y,test_size=0.1,train_size=0.9)


# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

In [29]:
X_train

Temperature °C  Mols KCL
549             200       182
154             813       887
11              307       781
587             996        52
421             408       731
..              ...       ...
190             600       874
26              539       725
296             997       478
262             655       886
951             121       528

[900 rows x 2 columns]

In [30]:
Y_train

549    7.614640e+04
154    1.474497e+06
11     5.006455e+05
587    1.156133e+05
421    6.166595e+05
           ...     
190    1.077825e+06
26     8.030359e+05
296    9.716241e+05
262    1.190948e+06
951    1.371933e+05
Name: Size nm^3, Length: 900, dtype: float64

In [61]:

reg = LinearRegression()
reg.fit(X_train,Y_train).predict(X_train)

array([-5.58184611e+04,  1.21490342e+06,  6.57798660e+05,  5.13929868e+05,
        6.95468645e+05,  9.47485236e+05, -2.21086418e+05,  3.22863795e+05,
        5.76175375e+05,  9.23305389e+05,  8.95595814e+05, -2.63164247e+05,
        5.63987154e+05,  6.54757782e+05,  4.19131768e+05,  4.87950877e+05,
        5.36296297e+05,  3.01848513e+05,  7.33970415e+05,  2.57829649e+05,
        5.13518374e+05,  9.15403290e+05, -1.86168568e+05,  8.13399910e+05,
        7.76484869e+05,  6.65696829e+05,  4.59627830e+05,  1.28374410e+05,
        3.61818428e+05,  5.16635674e+05,  1.33328952e+06,  8.72237410e+05,
        8.10681798e+05,  9.68745411e+05,  1.75432702e+05,  4.75890273e+05,
        1.69724907e+05,  1.11619274e+06,  1.00945716e+05, -1.42116197e+05,
        5.79212322e+05, -3.34752246e+05,  1.14016513e+06,  4.03011215e+05,
        8.86675159e+05,  4.87768553e+05,  1.33555528e+06, -2.93426880e+05,
        4.29985596e+05,  7.28349384e+05, -2.47083613e+05,  4.99267911e+05,
       -4.15923182e+05,  

In [58]:
reg.fit(X_train,Y_train).score(X_train,Y_train)

0.864116249429537

The Output score is known as the coeffiecient of determination, which is 1-u/v. This score seems to show how good its best fit line is by by showing improvments to the regression's cost by comparing its cost from its current values, vs its cost from its average. The max score it can have is 1.0 as it shows that the regression's cost is 0.

In [62]:
reg.coef_


array([ 884.5263742 , 1033.34357095])

In [64]:
reg.intercept_


-420792.26590027154

$$
  h(x)= [ 884.5263742 , 1033.34357095]*[x1,x2]^T-420792.26
$$


# Use the cross_val_score function to repeat your experiment across many shuffles of the data

# Report on their finding and their significance

In [88]:
scores = cross_val_score(reg, X, y, cv = 5)
scores

array([0.87218533, 0.84274655, 0.86144932, 0.8758663 , 0.86160014])

These scores represent the loss functions achieved from running the linear regression multiple times using various shuffles of data.

# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

# Report on the metrics and output the resultant equation as you did in Part 3.

In [120]:
poly = PolynomialFeatures(degree=2)
X_poly=poly.fit_transform(X)


In [127]:
Xpoly_train, Xpoly_test, Ypoly_train, Ypoly_test = train_test_split(X_poly, y, test_size=0.2)
pol = LinearRegression()
pol.fit(Xpoly_train,Ypoly_train).predict(Xpoly_train)


array([1.30749886e+06, 3.42182029e+05, 3.35545000e+05, 1.43166160e+06,
       1.46044903e+06, 5.83889114e+05, 4.55983857e+05, 4.82883143e+04,
       9.71624114e+05, 8.08596257e+05, 4.80946400e+05, 2.44549600e+05,
       1.23976740e+06, 9.38430829e+05, 5.55003429e+05, 1.41099457e+05,
       7.97577257e+05, 6.13645600e+05, 3.16833400e+05, 1.78553203e+06,
       5.13347114e+05, 5.97743114e+05, 5.62423143e+04, 6.40446429e+05,
       1.49123371e+06, 6.24742429e+05, 3.31453714e+05, 6.36024000e+04,
       5.38421457e+05, 5.47625257e+05, 7.95132257e+05, 3.74084457e+05,
       1.24615340e+06, 3.21153829e+05, 6.11333829e+05, 3.81815314e+05,
       1.02880114e+05, 5.45187114e+05, 1.15608846e+06, 2.35764114e+05,
       2.63859857e+05, 1.10952031e+06, 4.52268000e+05, 9.34857257e+05,
       1.02334711e+06, 8.14731430e+03, 1.61846203e+06, 1.27018346e+06,
       3.45304286e+04, 1.22792686e+06, 2.20722829e+05, 1.71829829e+05,
       4.59800314e+05, 5.56986600e+05, 5.37425257e+05, 4.26940257e+05,
      

In [128]:
pol.fit(Xpoly_train,Ypoly_train).score(Xpoly_train,Ypoly_train)

1.0

In [129]:
pol.coef_

array([ 0.00000000e+00,  1.20000000e+01, -1.52925930e-07, -3.92343935e-11,
        2.00000000e+00,  2.85714287e-02])

In [130]:
pol.intercept_


1.811108086258173e-05

$ h(x) =[ 0.00000000^0,  1.20000000^1, -1.52925930^{-7}, -3.92343935^{-11},2.00000000^0,  2.85714287^{-2}]*[1,x1,x2,x1^2,x1x2,x2^2]^T +1.811108086258173^{-5}$ 


This time the coeffiecient and the intercept values are much smaller this time. We also notice that the score is "perfect" when it comes to Polynomial regression